### PR TITLE
feat(detector): PLCOPEN-CP17: Usage of parameters shall match their declaration mode.

### PR DESCRIPTION
**Pull request by cangkui**
## PLCOPEN-CP17 

In PLCopen coding guideline document, PLCOPEN-CP17 is defined as follows:

**PLCOPEN-CP17:** Usage of parameters shall match their declaration mode.
**Description:** The parameters declared as input shall be read, the parameters declared as output shall be written and the parameters declared as in/out shall be read and written. 
**Guideline:** The mode of a parameter (input, output or input/output) should reflect the usage of the parameter in the corresponding POU. The following rules apply: 

- Each input parameter should be read at least once in the POU code 
- Each input parameter should not be written in the POU code 
- Each output parameter should be written at least once in the POU code 
- Each input/output parameter should be either read or written in the POU code 

**Reasoning:** Not using the parameters declared in the POU interface is a loss of time for the developers and for the machine performance. Using the parameter in the wrong way may lead to side effects when the Programming Support Environment doesn't limit the usage. 
**Exceptions:** None

## Comment & Explanation

This PR submitted a new detector for PLCOPEN-CP17 rule and a corresponding .st file as test case.

However, due to the implementation of current parser/CST definition, we got an explicit trouble in the new detector.
In IEC 61131-3, the formal definitions related to "Subscript Variable Access" are:
```ebnf
Symbolic_Variable  : ( ( 'THIS' '.' ) | Namespace_Hierarchy '.' )? ( Var_Access | Multi_Elem_Var );
Var_Access         : Variable_Name | Ref_Deref; 
Multi_Elem_Var     : Var_Access (Subscript_List | Struct_Variable)+;
Subscript_List     : '[' Subscript ( ',' Subscript )* ']';
Subscript          : Expression;
...
```
So theoretically, all variables in a `Subscript` should all be tagged to "read".
But currently the parser has discarded all variables in the `Subscript` inside the `[]` when it is not a constant, thus the detector cannot distinguish the read/write of the variables in the `Subscript`:
```mly
(* a code segment from parser.mly *)
let multi_elem_var :=
  | name_ti = var_access; T_LBRACK; sub_list = subscript_list; T_RBRACK;
  {
   let (name, ti) = name_ti in
   let sv = Syntax.SymVar.create name ti in
   (* Add array indexes to variable. *)
   List.fold_left
     sub_list
     ~init:sv
     ~f:(fun acc_sv e -> begin
         match e with
         | Syntax.ExprConstant (_,c) -> begin
           let val_opt = c_get_int c in
           match val_opt with
           | Some v -> Syntax.SymVar.add_array_index acc_sv v
           | None -> Syntax.SymVar.add_array_index_opaque acc_sv
         end
         | _ -> Syntax.SymVar.add_array_index_opaque acc_sv
     end)
  }
  (* | ~ = var_access; sub_list = nonempty_list(struct_variable); *)

let subscript_list :=
 | s = subscript;
 { [s] }
 | vs = subscript_list; T_COMMA; s = subscript;
 { List.append vs [s] }

let subscript :=
  | e = expression; <>
```

This is a question worth discussing.

### CODE REVIEW

File: src/lib/plcopen_cp17.ml [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    List.fold_left body ~init:acc ~f:collect_stmt
+  | S.StmWhile (_, cond, body) ->
+    let acc = collect_stmt (reads, writes) cond in
+    List.fold_left body ~init:acc ~f:collect_stmt
+  | S.StmRepeat (_, body, cond) ->
+    let acc = List.fold_left body ~init:(reads, writes) ~f:collect_stmt in
+    collect_stmt acc cond
+  | S.StmEmpty _ | S.StmExit _ | S.StmContinue _ | S.StmReturn _ ->
+    (reads, writes)
+  | S.StmFuncCall (_, _, params) ->
```
Comment (jubnzv) [inline src/lib/plcopen_cp17.ml:77]: I see the following problem here: `p.stmt` here is `formal := actual` where `formal` is the callee's parameter, not a variable of this POU -- so recursing into the generic `ASSIGN` branch incorrectly adds formal to the caller's write-set. The `=>` case already skips the LHS correctly; := needs the same treatment (walk only the rhs as a read).

Otherwise, it fires FPs in the current implementaiton:
```
FUNCTION_BLOCK Callee
VAR_INPUT
    InParam : INT;
END_VAR
VAR
    sink : INT;
END_VAR
    sink := InParam;
END_FUNCTION_BLOCK


FUNCTION_BLOCK Caller
VAR_INPUT
    InParam : INT;
END_VAR
VAR
    sub  : Callee;
    temp : INT;
END_VAR
    temp := InParam;
    sub(InParam := 42);
END_FUNCTION_BLOCK
```

**Review by jubnzv — CHANGES_REQUESTED** [review]
Thanks for the PR!

I see a FP in one corner case.

Could we fix it and add unit tests for it, including this case?

Additionally, is it supposed to support class methods?

File: src/lib/plcopen_cp17.ml (hunk 1 above)
Comment (cangkui) [reply src/lib/plcopen_cp17.ml:77]: @jubnzv Thanks a lot for reviewing and pointing this FP out. I have pushed a commit to fix it, with supplementing a unit test in test_plcopen.py.

For class, it seems that I have not considered about class methods in this detector yet, because i rarely used class in ST, which means I have no idea about how it works in IECClass. Maybe we can improve it in later commits/PRs.

Comment (jubnzv) [reply src/lib/plcopen_cp17.ml:77]: Thank you!
